### PR TITLE
Truncated iteration stats

### PR DESCRIPTION
**Pull request by snim2**
**NEEDS SQUASHING**

Latest PDF from bencher7 data: [b8_fixed.pdf](https://github.com/softdevteam/warmup_experiment/files/1199924/b8_fixed.pdf)

The logic for collecting stats for each truncated iterations JSON file are quite complicated. This is partly because the stats available for each <VM, benchmark> pair differ depending on the classification of the <VM, benchmark> (as per the tables). In particular:

  * If all pexecs are flat we DO NOT have data for steady iter (#, s)
  * If all pexecs are NSS we have no summary data
  * If the <VM, benchmark> is inconsistent and contains an NSS we have no summary data

Please check carefully!

### CODE REVIEW

**Comment by ltratt** [comment]
This looks like a good start! A couple of immediate comments.

There's something odd going on at about x=1930: the intersection is above the class line which, I think, should be impossible. Should we have an assertion that guarantees this can't happen?

The steady iter line and legend don't seem consistent? The legend has it as a dashed orange line, but the graph is orange circles with blue/orange dashed lines?

**Comment by snim2** [comment]
Ah, the second one is deliberate. The steady iter (#, s) lines coincide, so you are seeing the blue solid line underneath the dashed line.

**Comment by ltratt** [comment]
Oh they're always the same. Then we should probably just get rid of one of those lines, because they're both the same measure.

**Comment by snim2** [comment]
When we discussed this in the thrash, you took the view that they might be the same in our data and different in data from other machines. (To be clear, I'm not convinced of that, but it's probably worth thinking about)

**Comment by ltratt** [comment]
Not sure if we're talking about the same thing? Steady state (s) and steady iter (#) are the same measure by definition (just in different units).

**Comment by snim2** [comment]
We are. I'll get rid of one of the lines.

**Comment by snim2** [comment]
Test case (iteration 1970 was the problematic case):  [fix.pdf](https://github.com/softdevteam/warmup_experiment/files/1200405/fix.pdf)
 
debug output:

```
Truncated to 1900 iterations
summary[1900][steady_iter] = {u'different': 2, u'same': 25}, % same = 92.59
summary[1900][steady_state_time] = {u'different': 2, u'same': 25}, % same = 92.59
summary[1900][intersection] = {u'different': 10, u'same': 36}, % same = 78.26
summary[1900][classifications] = {u'different': 18, u'same': 1362}, % same = 98.70

Truncated to 1930 iterations
summary[1930][steady_iter] = {u'different': 1, u'same': 25}, % same = 96.15
summary[1930][steady_state_time] = {u'different': 1, u'same': 25}, % same = 96.15
summary[1930][intersection] = {u'different': 10, u'same': 36}, % same = 78.26
summary[1930][classifications] = {u'different': 18, u'same': 1362}, % same = 98.70

Truncated to 1970 iterations
summary[1970][steady_iter] = {u'different': 0, u'same': 26}, % same = 100.00
summary[1970][steady_state_time] = {u'different': 0, u'same': 26}, % same = 100.00
summary[1970][intersection] = {u'different': 6, u'same': 40}, % same = 86.96
summary[1970][classifications] = {u'different': 9, u'same': 1371}, % same = 99.35
```

**Comment by ltratt** [comment]
Looks plausible! Will be interesting to see the full graph.

**Comment by ltratt** [comment]
Although here's a question: I think that, at any point, the intersection line must be greater than or equal to the sum of the other 3 lines (in all cases taking 100-y as the measure, if that makes sense). But take x=1900. It looks to me like Class=3, steady iter=8, steady time=8, but the intersection is 22. That suggests that some pexecs are being classified as changed even though none of their classification, steady iter, or steady time doesn't match?

**Comment by snim2** [comment]
We are looking at each individual classification for each pexec, whereas we look at the steady time and steady iter for each <VM, benchmark> pair. So it may be that (for example) a "good inconsistent" <VM, benchmark> pair has similar steady iter and times, but different classifications in various pexecs.

**Comment by ltratt** [comment]
Erp, I've confused myself. Can you give me a concrete example of this to unstick my brain?

**Comment by snim2** [comment]
These are the tables for 1900 and 2000 iterations:

  * 1900: [trunc1900.pdf](https://github.com/softdevteam/warmup_experiment/files/1200723/trunc1900.pdf)
  * 2000: [trunc2000.pdf](https://github.com/softdevteam/warmup_experiment/files/1200727/trunc2000.pdf)

The summary for 1900 is:

```
summary[1900][steady_iter] = {u'different': 2, u'same': 25}, % same = 92.59
summary[1900][steady_state_time] = {u'different': 2, u'same': 25}, % same = 92.59
summary[1900][intersection] = {u'different': 10, u'same': 36}, % same = 78.26
summary[1900][classifications] = {u'different': 18, u'same': 1362}, % same = 98.70
```

So, I think your question is "why does `intersection[same] + intersection[different]` not equal `steady_iter[same] + steady_iter[different]`? 

The answer is that we always have classification data for all pexecs, but we don't always have steady iter and time data (e.g. when ALL pexecs are NSS). For example:

```
richards:C:default-c
process exec 21 differ in classification: no steady state flat
ALL bad inconsistent
```

So, in this case we would have had 1 classification `different`, 29 classifications `same` but no data for steady iter or time.

I've left the full debug output here: https://gist.github.com/snim2/85540f2ad11deb90288580435da99ca2 (but it is pretty unreadable).

**Comment by ltratt** [comment]
> The answer is that we always have classification data for all pexecs, but we don't always have steady iter and time data (e.g. when ALL pexecs are NSS).

Good point, I'd forgotten about that.

File: bin/plot_truncated_iterations [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+def do_iqrs_differ((x1, y1), (x2, y2)):
+    """Given two IQRs return True if they do NOT overlap."""
+
+    assert (y1 > x1 or x1 == y1) and (y2 > x2 or x2 == y2)
+    return y1 < x2 or y2 < x1
+
+
+def do_cis_differ(mean1, ci1, mean2, ci2):
+    """Given two CIs return True if they do NOT overlap."""
+
```
Comment (ltratt) [inline bin/plot_truncated_iterations]: If ci1 and cit2 are always positive (maybe it isn't?), the code below can be simplified to:

```
x1 = mean1 - ci1
y1 = mean1 + ci1
x2 = mean2 - ci2
y2 = mean2 + ci2
```

File: bin/plot_truncated_iterations [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+                summary['steady_state_time'][last_iter]['different'] += 1
+                summary['intersection'][last_iter]['different'] += 1
+                continue
+            # All three measures should be available in both the DEFAULT_ITER and last_iter cases.
+            if do_iqrs_differ(truncated_summaries[DEFAULT_ITERS]['machines'][machine][vm][bench]['steady_state_iteration_iqr'],
+                              truncated_summaries[last_iter]['machines'][machine][vm][bench]['steady_state_iteration_iqr']):
+                summary['steady_iter'][last_iter]['different'] += 1
+                intersection = False
+            else:
+                summary['steady_iter'][last_iter]['same'] += 1
```
Comment (ltratt) [inline bin/plot_truncated_iterations]: Should there be a `continue` here?

Comment (snim2) [reply bin/plot_truncated_iterations]: No, at this stage we have gone through all the special cases and we need to test the CIs for the steady iter and time values.

Comment (ltratt) [reply bin/plot_truncated_iterations]: OK.

File: bin/plot_truncated_iterations (hunk 1 above)
Comment (snim2) [reply bin/plot_truncated_iterations]: I ran through the small test case and didn't get any negative CIs, but in general they can be negative, so I'd be inclined to leave this.

Comment (ltratt) [reply bin/plot_truncated_iterations]: How can they be negative? If that does happen, it's probably a bug, and we'd probably end up printing CIs like "+/- -x" which would be... weird. So I'd suggest adding an `assert` that the CIs are `>= 0` and then simplify the code.

Comment (snim2) [reply bin/plot_truncated_iterations]: Ah, I forgot we do this with bootstrapping, so you are right.

**Comment by ltratt** [comment]
OK, so now I guess we wait for an updated graph then probably it'll be time for squashing.

**Comment by snim2** [comment]
Next PDF: [b8.pdf](https://github.com/softdevteam/warmup_experiment/files/1202317/b8.pdf)

**Comment by ltratt** [comment]
Interesting. I wonder why there's such a significant drop at x=1990?

**Comment by snim2** [comment]
Minor tweak: [b8_zorder.pdf](https://github.com/softdevteam/warmup_experiment/files/1202476/b8_zorder.pdf)

**Comment by ltratt** [comment]
Where are we in terms of tidying this up, reviewing it, and getting it merged?

**Comment by snim2** [comment]
[For the record] this PR is about to be squashed and force-pushed before re-review.

**Comment by ltratt** [comment]
Agreed.

**Comment by snim2** [comment]
Squashed and read for the next review.

File: bin/plot_truncated_iterations [hunk 3]
```diff
@@ -0,0 +1,374 @@
...
+import numpy
+import os
+import os.path
+import sys
+
+# We use a custom install of rpy2, relative to the top-level of the repo.
+sys.path.insert(0, os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(__file__))),
+                                'work', 'pylibs'))
+# R packages are stored relative to the top-level of the repo.
+os.environ['PATH'] = ':'.join([os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(__file__))),
```
Comment (ltratt) [inline bin/plot_truncated_iterations:56]: We know this stuff doesn't work, right?

File: bin/plot_truncated_iterations [hunk 4]
```diff
@@ -0,0 +1,374 @@
...
+    for key in original_results[machine]['classifications']:
+        if len(original_results[machine]['classifications'][key]) == 0:  # Skipped benchmark.
+            continue
+        class_counts = [original_results[machine]['classifications'][key].count(category) for category in CATEGORIES]
+        class_cis[key] = numpy.array(MCI.multinomialCI(rpy2.robjects.FloatVector(class_counts), ALPHA))
+    # Generate summary of results to be plotted.
+    summary = {'classifications': dict(), 'steady_iter': dict(), 'steady_state_time': dict(), 'intersection': dict()}
+    for last_iter in truncated_summaries:
+        if last_iter not in summary:
+            summary['classifications'][last_iter] = { 'same': 0, 'different': 0 }
```
Comment (ltratt) [inline bin/plot_truncated_iterations]: One thing I'd like to do in this function -- which is fairly easy, but will also help performance a fair bit -- is to change all the strings to constants e.g. `same` -> `SAME` (or whatever). In fact for the dictionaries which just have two keys, we'd be much better making them lists (e.g. `[<SAME>, <DIFFERENT>]`) which should also save a lot of memory (which we really need to do).

**Comment by ltratt** [comment]
OK so two things only. I think we should fix the `os.environ` thing in this PR: I don't mind if we fix the int constants & lists stuff in this PR or one that shortly follows it (provided it gets done soon).

File: bin/plot_truncated_iterations (hunk 3 above)
Comment (snim2) [reply bin/plot_truncated_iterations:56]: We do, but I would prefer to merge this and the other PR now and apply a fix for #430 to all the scripts that need it in one go.

Comment (ltratt) [reply bin/plot_truncated_iterations:56]: OK.

**Comment by snim2** [comment]
OK, I think this is ready for re-review / squashing...

**Comment by ltratt** [comment]
Can you give me a rough idea of the performance improvement from using lists instead of dicts?

File: bin/plot_truncated_iterations [hunk 5]
```diff
@@ -0,0 +1,376 @@
...
+    machine = original_results.keys()[0]
+    # Generate CIs for DEFAULT_ITER classification data.
+    class_cis = dict()
+    for key in original_results[machine]['classifications']:
+        if len(original_results[machine]['classifications'][key]) == 0:  # Skipped benchmark.
+            continue
+        class_counts = [original_results[machine]['classifications'][key].count(category) for category in CATEGORIES]
+        class_cis[key] = numpy.array(MCI.multinomialCI(rpy2.robjects.FloatVector(class_counts), ALPHA))
+    # Generate summary of results to be plotted.
+    summary = {'classifications': dict(), 'steady_iter': dict(), 'steady_state_time': dict(), 'intersection': dict()}
```
Comment (ltratt) [inline bin/plot_truncated_iterations]: This would also be better as a list.

**Comment by ltratt** [comment]
Ignore my performance comment for the time being -- let's get rid of more of these dictionaries-that-should-be-lists. We'll save a fair bit of memory and almost certainly make things quite a bit faster, both of which we need to improve!

**Comment by snim2** [comment]
With lists: 16mins, with dicts: 18mins (using test data, and not a quiet machine)

**Comment by ltratt** [comment]
OK, so that's a promising start, so let's rid ourselves of more dictionaries :)

**Comment by snim2** [comment]
OK, ready again for review...

**Comment by snim2** [comment]
Testing on a small dataset:

  * before changing dicts to lists (13 minues): [test_small_before_time.pdf](https://github.com/softdevteam/warmup_experiment/files/1314706/test_small_before_time.pdf)
  * after (12 minutes): [test_small_latest_time.pdf](https://github.com/softdevteam/warmup_experiment/files/1314709/test_small_latest_time.pdf)

**Comment by ltratt** [comment]
Ready to squash?

**Comment by snim2** [comment]
Splat.

**Review by ltratt — APPROVED** [review]

**Pull request merged**